src/Type/Type.php

```diff
@@ -0,0 +1,456 @@
+            internal: $internalType,
+            length: $length,
+            name: InternalType::UNKNOWN === $internalType ? $userType : null,
+            precision: $precision,
+            scale: $scale,
+            unsigned: $unsigned,
+            withTimeZone: $withTimeZone,
+        );
+    }
+
+    /**
+     * Get arbitrary key based upon type name.
+     *
+     * @internal
+     *   For converter caching only.
+     */
+    public function getArbitraryName(): string
+    {
+        return match ($this->internal) {
+            InternalType::BINARY => 'binary',
+            InternalType::BOOL => 'bool',
+            InternalType::CHAR => 'char',
+            InternalType::DATE => 'date',
+            InternalType::DATE_INTERVAL => 'date_interval',
+            InternalType::DECIMAL => 'decimal',
+            InternalType::FLOAT => 'float',
+            InternalType::FLOAT_BIG => 'float_big',
+            InternalType::FLOAT_SMALL => 'float_small',
+            InternalType::IDENTITY => 'id',
+            InternalType::IDENTITY_BIG => 'id_big',
+            InternalType::IDENTITY_SMALL => 'id_small',
+            InternalType::INT => 'int',
+            InternalType::INT_BIG => 'int_big',
+            InternalType::INT_SMALL => 'int_small',
+            InternalType::JSON => 'json',
+            InternalType::NULL => 'null',
+            InternalType::SERIAL => 'serial',
+            InternalType::SERIAL_BIG => 'serial_big',
+            InternalType::SERIAL_SMALL => 'serial_small',
+            InternalType::TEXT => 'text',
+            InternalType::TIME => 'time',
+            InternalType::TIMESTAMP => 'timestamp',
+            InternalType::VARCHAR => 'varchar',
+            InternalType::UNKNOWN => $this->name ?? new QueryBuilderError("Unhandled types must have a type name."),
+            InternalType::UUID => 'uuid',
+        };
+    }
+
+    public function isDate(): bool
+    {
+        return $this->isDate ??= $this->internal->isDate();
+    }
+
+    public function isNumeric(): bool
+    {
+        return $this->isNumeric ??= $this->internal->isNumeric();
+    }
+
+    public function isText(): bool
+    {
+        return $this->isText ??= $this->internal->isText();
+    }
+
+    public function cleanUp(): self
+    {
+        return new self(
+            array: $this->array,
+            internal: $this->internal,
+            length: $this->internal->lengthIsRelevant() ? $this->length : null,
+            name: $this->internal === InternalType::UNKNOWN ? $this->name : null,
+            precision: $this->internal->precisionIsRelevant() ? $this->precision : null,
+            scale: $this->internal->scaleIsRelevant() ? $this->scale : null,
+            unsigned: $this->unsigned,
+            withTimeZone: $this->withTimeZone,
+        );
+    }
+
+    public function toArray(): self
+    {
+        $ret = clone $this;
+        $ret->array = true;
+
+        return $ret;
+    }
+
+    public function toNoArray(): self
```

Comment:
```suggestion
    public function toNonArray(): self
```

Reply:
Good idea, I was really unhappy with this name, but I can't simply apply your suggestion I need to fix core everywhere it's called.

Reply:
Done !